### PR TITLE
i16601 Fix class name in error message

### PR DESCRIPTION
**Pull request by nizhikov**

### CODE REVIEW

**Comment by nizhikov** [comment]
fixes #16601

**Review by dwijnand — DISMISSED** [review]

**Comment by nizhikov** [comment]
Hello @dwijnand - thanks for the review.
I replaced `asInstanceOf` with `match`.
Can you, please, advice me - what compiler suppose to do in case type not expected?
Right now I fallback to previous message.

```scala
        if (cls.isOneOf(AbstractOrTrait)) {
          srcTp.underlyingClassRef(refinementOK = false) match 
            case tref0: TypeRef => 
              val srcCls = tref0.symbol
              report.error(CantInstantiateAbstractClassOrTrait(srcCls, isTrait = srcCls.is(Trait)), pos)
            case _ => 
              report.error(CantInstantiateAbstractClassOrTrait(cls, isTrait = cls.is(Trait)), pos)
        }
```

File: compiler/src/dotty/tools/dotc/typer/Checking.scala [hunk 1]
```diff
@@ -185,12 +185,18 @@ object Checking {
...
    */
-  def checkInstantiable(tp: Type, pos: SrcPos)(using Context): Unit =
+  def checkInstantiable(tp: Type, srcTp: Type, pos: SrcPos)(using Context): Unit =
     tp.underlyingClassRef(refinementOK = false) match
       case tref: TypeRef =>
         val cls = tref.symbol
-        if (cls.isOneOf(AbstractOrTrait))
-          report.error(CantInstantiateAbstractClassOrTrait(cls, isTrait = cls.is(Trait)), pos)
+        if (cls.isOneOf(AbstractOrTrait)) {
+          srcTp.underlyingClassRef(refinementOK = false) match
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala:193]: It seems you could use `srcTp.underlyingClassRef(refinementOK = false).typeSymbol.is(Trait)` .

Comment (nizhikov) [reply compiler/src/dotty/tools/dotc/typer/Checking.scala]: Yes. Thanks.
With `srcTp.underlyingClassRef(refinementOK = false).typeSymbol` I can get rid of pattern matching and seems all tests still pass.

Please, take a look at pushed changes.

**Review by dwijnand — APPROVED** [review]

**Comment by nizhikov** [comment]
@dwijnand Thanks for the review.

All checks passed. Anything else should be done prior merge?

**Pull request merged**